### PR TITLE
For #21551 - Add delete history metadata in the History view

### LINKED ISSUE #21551: Add delete history metadata in History view

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #21551
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt [hunk 1]
```diff
@@ -37,6 +37,11 @@ class DefaultPagedHistoryProvider(
 
     private var historyGroups: List<History.Group>? = null
 
+    /**
+     * Map of [History.Metadata] to their corresponding [History.Regular].
+     */
+    private var metadataToVisitMap = mutableMapOf<History.Metadata, History.Regular>()
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt:43]: There is a concurrency issue here. The `getHistory` call on the provider is invoked on a background thread, and if multiple threads call it concurrently they may read the wrong data (from a wrong provider or wrong `getHistory` call). We'd have to use a different key (incl. offset/numberOfItems) and use a concurrent map, if we're keeping this approach.

File: app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt [hunk 2]
```diff
@@ -129,9 +142,17 @@ class DefaultPagedHistoryProvider(
         }
         val historyMetadata = historyGroupsInOffset.flatMap { it.items }
 
-        // Add all items that are not in a group filtering out any matches with a history metadata
-        // item.
-        result.addAll(history.filter { item -> historyMetadata.find { it.url == item.url } == null })
+        // Add all history items that are not in a group filtering out any matches with a history
+        // metadata item.
+        for (historyItem in history) {
+            val matchedMetadata = historyMetadata.find { it.url == historyItem.url }
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt:148]: This matching logic isn't sufficient yet. You could have `n` visits to the same page, and they are returned in desc order by visit timestamp. So you're basically always saving the oldest visit in the map currently which would then get deleted.

I think what we need is similar to the logic we already have, a mapping based on `visitedAt` and `createdAt` (for metadata). So for every metadata record we delete we will need to look for the corresponding visit and delete it too.

File: app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt (hunk 1 above)
Comment (csadilek) [reply app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt:43]: See my comment below though about mapping. I think we should not use a map here, but delete the closest visit by time.

File: app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt [hunk 3]
```diff
@@ -90,6 +91,31 @@ class DefaultPagedHistoryProvider(
...
+                VisitType.DOWNLOAD,
+                VisitType.REDIRECT_TEMPORARY,
+                VisitType.RELOAD,
+                VisitType.EMBED,
+                VisitType.FRAMED_LINK,
+                VisitType.REDIRECT_PERMANENT
+            )
+        )
+        return history.lastOrNull { it.url == historyMetadata.url }.also {
+            historyGroups = null
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt:115]: So, that's technically in the wrong place. it should be right after you call `delete` for the group itself.

I think in practice it'll work out but it's pretty weird having this reset here in the lookup. I.e. this simple lookup function doesn't know anything about group removals, why should it reset the state.

**Review by grigoryk — APPROVED** [review]
Just one tweak to where we reset `historyGroups` and I think this is good!

**Review by csadilek — APPROVED** [review]
🚢

**Pull request merged**